Support compiling qgis_analysis when QT_NO_PRINTER is defined.

OlivierLDff opened the pull request:
This commit allow compilation for iOS target where QtPrinter isn't available.
Affected classes:
* QgsLayoutAtlasToImageAlgorithm
* QgsLayoutAtlasToPdfAlgorithm
* QgsLayoutToImageAlgorithm
* QgsLayoutToPdfAlgorithm

**Related links**
https://forum.qt.io/topic/107731/printsupport-missing-on-ios
https://bugreports.qt.io/browse/QTBUG-34555

nyalldawson commented:
Instead of.having non functional algorithms, I'd rather the whole algorithm is not compiled not available. (Ie move the logic to cmake). Can you do this instead please?

OlivierLDff commented:
I was following how it have been done here:
https://github.com/qgis/QGIS/blob/9e02078d8801c9c125119f11b17da0b41a73b2d5/src/core/layout/qgslayoutexporter.h#L29

But yes disabling them from cmake would be much cleaner.
So should i append to `QGIS_ANALYSIS_SRCS` processing/qgsalgorithmlayouttopdf.cpp, ... if QT_PRINTER is there. Would it be ok for you?


I'm not that familiar with QGIS code base. Won't it break anywhere?
Where is for `qgsalgorithmlayoutatlastoimage.h` included?

From what is see i only need to add some ifndef in `analysis/processing/qgsnativealgorithms.cpp` for example : https://github.com/qgis/QGIS/blob/d45790e05b6cac71b5d5715187bde7171a0bd50b/src/analysis/processing/qgsnativealgorithms.cpp#L94

nyalldawson commented:
@OlivierLDff that sounds like a perfect approach! And yes, you'd conditionally include the algorithm in qgsnativealgorithms just like you described

OlivierLDff committed:
Fix qgis_analysis build when Qt::PrintSupport isn't available (ie iOs)

Disable:
- QgsLayoutAtlasToImageAlgorithm
- QgsLayoutAtlasToPdfAlgorithm
- QgsLayoutToImageAlgorithm
- QgsLayoutToPdfAlgorithm

OlivierLDff commented:
This should be the one.
I used a `if(TARGET Qt5::PrintSupport)` to know if Printer is there. I think it is cleaner than a if(IOS).
Note that to ensure compatibility with Qt6, it should be interesting to use [versionless tagging](https://doc-snapshots.qt.io/qt6-dev/cmake-qt5-and-qt6-compatibility.html#versionless-targets). But i don't know what is minimum version of Qt for QGIS, but i doubt that your minimal version of Qt is 5.15..

nyalldawson commented:
Looks good!

For completeness, you should also introduce the `#ifndef QT_NO_PRINTER ` into testqgsprocessingalgs.cpp so that the tests reliant on these algorithms aren't run if the algorithms aren't being compiled. (Otherwise you'll always get test failures on your build)

OlivierLDff commented:
I see that some test fail, i don't think it is my fault, i see other PR having the same error.
